apps/src/JSInterpreter.js

```diff
@@ -668,6 +672,36 @@ JSInterpreter.prototype.getProperty = function (
 };
 
 /**
+ * Wrapper to Interpreter's hasProperty (extended for custom marshaling)
+ *
+ * Does the named property exist on a data object.
+ * @param {!Object} interpeter Interpreter instance.
+ * @param {!Function} baseHasProperty Original hasProperty() implementation.
+ * @param {!Object} obj Data object.
+ * @param {*} name Name of property.
+ * @return {boolean} True if property exists.
+ */
+JSInterpreter.prototype.hasProperty = function (
+    interpreter,
+    baseHasProperty,
+    obj,
+    name) {
+  name = name.toString();
+  var nativeParent;
+  if (obj.isCustomMarshal ||
+      (obj === this.globalScope &&
+          (!!(nativeParent = this.customMarshalGlobalProperties[name])))) {
```

Comment:
There's a fair amount of duplication between this method and `getProperty`, particularly this convoluted boolean expression containing an assignment, and the check below for whether to use `obj` or `nativeParent`.  Doing an extraction here might both deduplicate code and make this easier to read.

If I understand correctly, you might end up with something like this:
```javascript
JSInterpreter.prototype.getPropertyNativeParent_ = function (obj, name) {
  if (obj.isCustomMarshal) {
    return obj.data;
  } else if (obj === this.globalScope) {
    return this.customMarshalGlobalProperties[name];
  }
  return undefined;
};

JSInterpreter.prototype.hasProperty = function (...) {
  name = name.toString();
  var nativeParent = this.getPropertyNativeParent(obj, name);
  if (nativeParent) {
    return name in nativeParent;
  } else {
    return baseHasProperty.call(interpreter, obj, name);
  }
};
```